### PR TITLE
MGDOBR-1031: Investigate monitoring and alerting strategy for the Data Plane

### PR DESCRIPTION
**Pull request by manstis**
**Export metrics from Operator**

I also refactored some code to better support re-use across `manager` and `operator`.

[https://issues.redhat.com/browse/MGDOBR-1031](https://issues.redhat.com/browse/MGDOBR-1031)

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] The layers in the `kustomize` folder have been updated accordingly.
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

File: manager/src/main/java/com/redhat/service/smartevents/manager/metrics/ManagedResourceOperationMapper.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
+package com.redhat.service.smartevents.manager.metrics;
+
+import java.util.Objects;
+
+import com.redhat.service.smartevents.infra.metrics.MetricsOperation;
+import com.redhat.service.smartevents.infra.models.dto.ManagedResourceStatus;
+import com.redhat.service.smartevents.infra.models.dto.ManagedResourceStatusUpdateDTO;
+import com.redhat.service.smartevents.manager.models.ManagedResource;
+
+public final class ManagedResourceOperationMapper {
```
Comment (manstis) [inline manager/src/main/java/com/redhat/service/smartevents/manager/metrics/ManagedResourceOperationMapper.java:10]: We used to have some (IMO, horrible) code in both `BridgeService` and `ProcessorService` that decided what _operation_ had been performed and therefore what _action_ to complete. I moved it to this new class and refactored it. IMO it is now nicer.. but this is obviously very subjective :smile:

File: shard-operator/src/test/java/com/redhat/service/smartevents/shard/operator/BridgeExecutorServiceTest.java [hunk 2]
```diff
@@ -92,7 +92,6 @@ public void setup() {
         // point of failed completely. There is therefore a good chance there's an incomplete BridgeExecutor
         // in k8s when a subsequent test starts. This leads to non-deterministic behaviour of tests.
         // This ensures each test has a "clean" k8s environment.
-        kubernetesClient.resources(BridgeExecutor.class).inAnyNamespace().delete();
```
Comment (manstis) [inline shard-operator/src/test/java/com/redhat/service/smartevents/shard/operator/BridgeExecutorServiceTest.java:95]: This was already being done in the cleanup code.

File: shard-operator/src/test/java/com/redhat/service/smartevents/shard/operator/BridgeIngressServiceTest.java [hunk 3]
```diff
@@ -88,7 +88,6 @@ public void setup() {
         // point of failed completely. There is therefore a good chance there's an incomplete BridgeIngress
         // in k8s when a subsequent test starts. This leads to non-deterministic behaviour of tests.
         // This ensures each test has a "clean" k8s environment.
-        kubernetesClient.resources(BridgeIngress.class).inAnyNamespace().delete();
```
Comment (manstis) [inline shard-operator/src/test/java/com/redhat/service/smartevents/shard/operator/BridgeIngressServiceTest.java:91]: This was already being done in the cleanup code.

File: infra/src/main/java/com/redhat/service/smartevents/infra/metrics/MetricsOperation.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package com.redhat.service.smartevents.infra.metrics;
+
+import io.micrometer.core.instrument.Tag;
+
+/*
+    Constants for the logical user API operations we are report SLI data for.
+ */
+public enum MetricsOperation {
```
Comment (manstis) [inline infra/src/main/java/com/redhat/service/smartevents/infra/metrics/MetricsOperation.java:8]: I moved this to `infra`. It contains _ALL_ operations for which we emit metrics.

It used to be in just `manager` and an equivalent was in `operator`. I've combined and moved.

File: infra/src/main/java/com/redhat/service/smartevents/infra/metrics/MetricsService.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package com.redhat.service.smartevents.infra.metrics;
+
+import java.util.Set;
+
+public interface MetricsService<T> {
```
Comment (manstis) [inline infra/src/main/java/com/redhat/service/smartevents/infra/metrics/MetricsService.java:5]: I made the `MetricsService` in both `manager` and `operator` implement a common interface.

Previously we had two implementations that had different APIs and usage.

**Comment by afalhambra** [comment]
re-triggering manager CI job as it is failing

**Comment by afalhambra** [comment]
@manstis - failed tests are not related to you PR. I can see a bad request in the logs for the E2E:

```
  "aws_access_key": {
      
  },
  "aws_secret_key": {
      
  },
```
It seems to me that it's not picking up latest changes from upstream. There was a new secret added in GH, and some CI jobs were updated... - https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/1061/

**Comment by afalhambra** [comment]
Manager CI is failing due to DLQ E2E test merged yesterday, checking that issue further, but it shouldn't be related to your PR @manstis

**Comment by afalhambra** [comment]
@manstis I've just removed the flaky test [1] for the time being till I find out what is going on... please rebase and try again

https://github.com/5733d9e2be6485d52ffa08870cabdee0/sandbox/pull/1168

File: shard-operator/src/test/java/com/redhat/service/smartevents/shard/operator/ManagerSyncServiceTest.java [hunk 6]
```diff
@@ -231,4 +244,45 @@ public void testProcessorsAreDeletedWhenNotDeployed() throws JsonProcessingExcep
...
+                .contentType(ContentType.JSON)
+                .when()
+                .get("/q/metrics")
+                .then()
+                .extract()
+                .body()
+                .asString();
+
+        // Not all metrics are recorded by this test
+        assertThat(metrics).contains(managerRequestsTotalCountMetricName);
```
Comment (jiripetrlik) [inline shard-operator/src/test/java/com/redhat/service/smartevents/shard/operator/ManagerSyncServiceTest.java:263]: What is the reason for not covering all the metrics?

Comment (manstis) [reply shard-operator/src/test/java/com/redhat/service/smartevents/shard/operator/ManagerSyncServiceTest.java:263]: The test only handles successful creation; but one metric relates to failures.

Comment (jiripetrlik) [reply shard-operator/src/test/java/com/redhat/service/smartevents/shard/operator/ManagerSyncServiceTest.java:263]: Would it be also possible to add some test which simulates failure?

Comment (manstis) [reply shard-operator/src/test/java/com/redhat/service/smartevents/shard/operator/ManagerSyncServiceTest.java:292]: Done.

**Review by jiripetrlik — APPROVED** [review]
Thank you @manstis for adding a negative test scenario. Looks good to me.

**Review by lucamolteni — APPROVED** [review]

**Pull request merged**